Secvertex amvf to be merged into main EICrecon repo

bishoyDHD opened the pull request:
### Briefly, what does this PR introduce?
This adds secondary vertex finder to the EICrecon.

### What kind of change does this PR introduce?
- [ ] Bug fix (issue #__)
- [ ] New feature (issue #__)
- [ ] Documentation update
- [ ] Other: __

### Please check if this PR fulfills the following:
- [ ] Tests for the changes have been added
- [ ] Documentation has been added / updated
- [ ] Changes have been communicated to collaborators

### Does this PR introduce breaking changes? What changes might users need to make to their code?
No need to make changes to existing code. Full compliance with main EICrecon repo has been maintained throughout

### Does this PR change default behavior?
No change to default behaviour

bishoyDHD committed:
updating secondary vertex finder

bishoyDHD committed:
Merge branch 'eic:main' into displacedvertex

bishoyDHD committed:
added reco-particles to displaced vertex

bishoyDHD committed:
Merge branch 'displacedvertex' of https://github.com/bishoyDHD/EICrecon into displacedvertex

bishoyDHD committed:
Prelim secvertex finder

bishoyDHD committed:
Modifying the displaced vertex finder

bishoyDHD committed:
adding displaced vertex changes

bishoyDHD committed:
Merge branch 'eic:main' into displacedvertex

bishoyDHD committed:
Added more functionality to sec. vertex and now compatible with ACTS >= 33

bishoyDHD committed:
Merge branch 'displacedvertex' of https://github.com/bishoyDHD/EICrecon into displacedvertex

bishoyDHD committed:
Merge branch 'eic:main' into displacedvertex

bishoyDHD committed:
adding fuctions for sec vertex reco

bishoyDHD committed:
Merge branch 'eic:main' into displacedvertex

bishoyDHD committed:
Secondary Vertexing Utility tool implentation

bishoyDHD committed:
Updating sec vertex finder

bishoyDHD committed:
Merge branch 'eic:main' into displacedvertex

bishoyDHD committed:
Merge branch 'eic:main' into displacedvertex

bishoyDHD committed:
Added method to calculate the Prim. Vtx form Adaptive Multivertex Finder

bishoyDHD committed:
Updated the SecondaryVertex to calculate prim. vtx using AMVF

bishoyDHD committed:
cleaning up sec. vertex finder

bishoyDHD committed:
Cleaned up some std::cout printouts

bishoyDHD committed:
Merge branch 'eic:main' into displacedvertex

bishoyDHD committed:
added C macro for Lambda->proton pi analysis

bishoyDHD committed:
Tidying up plots for Lambda decay

Commit:
Added reco Lambda ROOT file

bishoyDHD committed:
fixed some minor bugs in Lambda analysis macro

Commit:
updating Lambda reco ROOT file

bishoyDHD committed:
added runlist here

bishoyDHD committed:
Added 4th tracking dimension of time. Working properly

bishoyDHD committed:
added sec. vertex invariant mass

bishoyDHD committed:
Added more functionality to sec. verting analysis Macro

Commit:
Newly generate ROOT for plotting

bishoyDHD committed:
added Armenteros Podolansky plot

bishoyDHD committed:
Updated the ana file

bishoyDHD committed:
sec vertex cuts

bishoyDHD committed:
Merge branch 'eic:main' into displacedvertex

bishoyDHD committed:
Fixing sec vertex analysis

bishoyDHD committed:
Updating sec vertex to compliance with new changes

bishoyDHD committed:
Complaince update with main branch in recent singularity updates

bishoyDHD committed:
Merge branch 'eic:main' into displacedvertex

bishoyDHD committed:
Added AMVPrimaryVertex Collection

bishoyDHD committed:
Merge branch 'displacedvertex' of https://github.com/bishoyDHD/EICrecon into displacedvertex

bishoyDHD committed:
Merge branch 'eic:main' into displacedvertex

bishoyDHD committed:
Modified AMVF algorithm for both primary and sec vertexting

bishoyDHD committed:
Update Sec vertex, field-->bField

bishoyDHD committed:
Update sec. vertex to be in compliance with updated ACTS

bishoyDHD committed:
Playing with AMVF for PV

bishoyDHD committed:
small updates to Sec. Vertex finder

bishoyDHD committed:
Merge branch 'eic:main' into displacedvertex

bishoyDHD committed:
cleaning up displaced vertex

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

bishoyDHD committed:
Displacedvertex (#1912)

### Briefly, what does this PR introduce?


### What kind of change does this PR introduce?
- [ ] Bug fix (issue #__)
- [ ] New feature (issue #__)
- [ ] Documentation update
- [ ] Other: __

### Please check if this PR fulfills the following:
- [ ] Tests for the changes have been added
- [ ] Documentation has been added / updated
- [ ] Changes have been communicated to collaborators

### Does this PR introduce breaking changes? What changes might users
need to make to their code?

### Does this PR change default behavior?

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: Dmitry Kalinkin <dmitry.kalinkin@gmail.com>
Co-authored-by: Simon Gardner <simon.gardner@glasgow.ac.uk>
Co-authored-by: Tommy Chun Yuen Tsang <tommytsang3108@hotmail.com.hk>
Co-authored-by: Souvik Paul <136603463+eicsouvik@users.noreply.github.com>
Co-authored-by: Desktop ssedd1123 <ssedd1123@DESKTOP-0J5IETB>
Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: Derek M Anderson <derek.murphy.anderson@protonmail.com>
Co-authored-by: Sebouh Paul <sebouh.paul@gmail.com>
Co-authored-by: Nathan Brei <nathan.w.brei@gmail.com>
Co-authored-by: Alex Jentsch <alexmichaeljentsch@gmail.com>
Co-authored-by: Alexander Jentsch <ajentsch@bnl.gov>
Co-authored-by: Wouter Deconinck <wdconinc@gmail.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Dongwi H Dongwi <ddongwi@eic0102.rcf.bnl.gov>
Co-authored-by: Tristan Protzman <tlprotzman@gmail.com>
Co-authored-by: EIC Container Build Service <gitlab@eicweb.phy.anl.gov>
Co-authored-by: Barak Schmookler <bschmookler1@gmail.com>
Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>
Co-authored-by: minho.kim@anl.gov <minho.kim@anl.gov>
Co-authored-by: Sylvester Joosten <sylvester.joosten@gmail.com>
Co-authored-by: Sylvester Joosten <sjoosten@anl.gov>

bishoyDHD committed:
fixed spelling mistakes in comments eicrecon -Pnthreads=8 -Ptracking:CentralTrackSeedingResults:impactMax=15.0 ../simfiles/sim_ab.eAu_106.0.edm4hep.root -Ppodio:output_file=testingLambdaC.root eicrecon -Pnthreads=8 -Ptracking:CentralTrackSeedingResults:impactMax=15.0 ../simfiles/sim_ab.eAu_106.0.edm4hep.root -Ppodio:output_file=testingLambdaC.root

bishoyDHD committed:
Merge branch 'displacedvertex' of https://github.com/bishoyDHD/EICrecon into displacedvertex

bishoyDHD committed:
bringing branch into compliance with updated EICrecon

bishoyDHD committed:
merging some conflicts here

wdconinc commented:
Fails to build. We typically expect checks to complete before we do a review. Initial comments: I don't think we will want to add TProfile dependencies.

bishoyDHD committed:
removed the TProfile dependcies

bishoyDHD committed:
removed unused variables because they cause errors

bishoyDHD committed:
removing even more unused variables

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

bishoyDHD committed:
found more unused vars and removed them for now

bishoyDHD committed:
Found additional unused vars

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

bishoyDHD committed:
found yet another unused variable

bishoyDHD committed:
Merge branch 'secvertexAMVF' of https://github.com/eic/EICrecon into displacedvertex

bishoyDHD committed:
cleaning up sec vertex finder

bishoyDHD committed:
removed more unused vars from SecondaryVertexFinder_factory

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

bishoyDHD committed:
Changed to the correct type in the SecondaryVertexFinder_factory

bishoyDHD committed:
secvertexAMVF compliance for type matching

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

bishoyDHD committed:
fixing TrackingUtilityTool for sec vertex finder

bishoyDHD committed:
Merge branch 'secvertexAMVF' of https://github.com/eic/EICrecon into displacedvertex

bishoyDHD committed:
Fixing Clang cleanup

bishoyDHD committed:
Removed UtilityTool for now

bishoyDHD committed:
Cleaning up the preprocessor include directives

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

veprbl reviewed (commented):
(Not a review)

bishoyDHD committed:
Updated License and Copyright statements: Dmitry

bishoyDHD committed:
Merge branch 'secvertexAMVF' of https://github.com/eic/EICrecon into displacedvertex

bishoyDHD committed:
Updated missing copyright

bishoyDHD commented on src/factories/tracking/SecondaryVertexFinder_factory.h:
```diff
@@ -0,0 +1,60 @@
+// Created by Joe Osborn
+// Subject to the terms in the LICENSE file found in the top-level directory.
+//
```
For completeness, I have made the aforementioned suggestions. Still fails for checks beyond my control

wdconinc commented on src/algorithms/tracking/SecondaryVertexFinder.h:
```diff
@@ -0,0 +1,338 @@
+                       vtx.fullCovariance()(1, 3), vtx.fullCovariance()(2, 3));
+    auto eicvertex = prmVertices->create();
+    eicvertex.setType(1); // boolean flag if vertex is primary vertex of event
+    eicvertex.setChi2((float)vtx.fitQuality().first); // chi2
+    eicvertex.setNdf((float)vtx.fitQuality().second); // ndf
+    eicvertex.setPosition({
+        (float)vtx.position().x(),
+        (float)vtx.position().y(),
+        (float)vtx.position().z(),
+        (float)vtx.time(),
```
Avoid C-style casts.

wdconinc commented on src/algorithms/tracking/SecondaryVertexFinder.h:
```diff
@@ -0,0 +1,338 @@
+          const auto& traj    = trk.getTrajectory();
+          const auto& trkPars = traj.getTrackParameters();
+          for (const auto& par : trkPars) {
+            const double EPSILON = 1.0e-4; // mm
```
Use proper `std::numeric_limits::epsilon`.

wdconinc commented on src/algorithms/tracking/SecondaryVertexFinder.h:
```diff
@@ -0,0 +1,338 @@
+                           secvertex.fullCovariance()(1, 3), secvertex.fullCovariance()(2, 3));
+        auto eicvertex = secVertices->create();
+        eicvertex.setType(0); // boolean flag if vertex is primary vertex of event
+        eicvertex.setChi2((float)secvertex.fitQuality().first); // chi2
+        eicvertex.setNdf((float)secvertex.fitQuality().second); // ndf
+        eicvertex.setPosition({
+            (float)secvertex.position().x(),
+            (float)secvertex.position().y(),
+            (float)secvertex.position().z(),
+            (float)secvertex.time(),
```
Avoid C-style casts.

wdconinc commented on src/algorithms/tracking/SecondaryVertexFinder.h:
```diff
@@ -0,0 +1,338 @@
+              const auto& traj    = trk.getTrajectory();
+              const auto& trkPars = traj.getTrackParameters();
+              for (const auto& par : trkPars) {
+                const double EPSILON = 1.0e-4; // mm
```
Use `std::numeric_limits::epsilon`.

wdconinc commented on src/algorithms/tracking/SecondaryVertexFinder.h:
```diff
@@ -0,0 +1,338 @@
+              const auto& trkPars = traj.getTrackParameters();
+              for (const auto& par : trkPars) {
+                const double EPSILON = 1.0e-4; // mm
+                if (fabs((par.getLoc().a / edm4eic::unit::mm) - (loc_a / Acts::UnitConstants::mm)) <
+                        EPSILON &&
+                    fabs((par.getLoc().b / edm4eic::unit::mm) - (loc_b / Acts::UnitConstants::mm)) <
+                        EPSILON) {
```
Avoid `fabs`, use `std::abs`. (also earlier)

wdconinc commented on src/factories/tracking/SecondaryVertexFinder_factory.h:
```diff
@@ -0,0 +1,59 @@
+  void ChangeRun(int32_t /*run_number*/) {}
+
+  void Process(int32_t /*run_number*/, uint64_t /*event_number*/) {
+    std::tie(prm_vertices_output(), sec_vertices_output()) =
+        m_algo->produce(m_reco_input(), m_acts_trajectories_input());
```
Process will be `void`, so needs two tuples by reference, input and output.

bishoyDHD replied:
In this case `std::tie` already takes cares of this, and is able to output `std::tuple` which is then written to the output `ROOT file`. Otherwise this would fail at `runtime`.

wdconinc replied:
No, the function signature will be different when you use the algorithms interface as requested elsewhere.

wdconinc commented on src/factories/tracking/SecondaryVertexFinder_factory.h:
```diff
@@ -0,0 +1,59 @@
+  void Process(int32_t /*run_number*/, uint64_t /*event_number*/) {
+    std::tie(prm_vertices_output(), sec_vertices_output()) =
+        m_algo->produce(m_reco_input(), m_acts_trajectories_input());
+    /*
+    sec_vertices_output() = m_algo->produce(m_reco_input(),m_acts_trajectories_input());
+    */
```
Remove.

wdconinc commented on src/global/tracking/tracking.cc:
```diff
@@ -214,6 +215,14 @@ void InitPlugin(JApplication* app) {
+  app->Add(new JOmniFactoryGeneratorT<SecondaryVertexFinder_factory>(
+      "AMVSecondaryTrackVertices", {"ReconstructedParticles", "CentralCKFActsTrajectories"},
+      {
+          "AMVPrimaryVertices",
+          "AMVSecondaryTrackVertices",
```
Don't prefix with name of algorithm. If anything, suffix. Be consistent and don't call something `PrimaryVertices` but the other thing `SecondaryTrackVertices`. Use Track or not, but consistently.

wdconinc reviewed (commented):
Main comment for discussion: name of output collection. I'd suggest `SecondaryVertices`.

Rest is style and maintainability.

bishoyDHD committed:
Cleanups. Fixed C-style cast

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

bishoyDHD committed:
Cleaning up SecondaryVertexFinder.h

bishoyDHD committed:
Merge branch 'secvertexAMVF' of https://github.com/eic/EICrecon into displacedvertex

bishoyDHD committed:
Revert back to eicrecon::WithPodConfig

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

bishoyDHD committed:
iwyu: SecondaryVertexFinder

bishoyDHD committed:
Merge branch 'secvertexAMVF' of https://github.com/eic/EICrecon into displacedvertex

bishoyDHD committed:
Using  interface. Moved variables to config. Renamed collection names in the Factories

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

bishoyDHD commented on src/algorithms/tracking/SecondaryVertexFinder.h:
```diff
@@ -0,0 +1,338 @@
+
+namespace eicrecon {
+class SecondaryVertexFinder
+    : public eicrecon::WithPodConfig<eicrecon::SecondaryVertexFinderConfig> {
```
After pushing `git precomp running` produces the following error at bottom of this message(non my work machine and also BNL):

For the new algorithms, the `init(){}` is a default function. However, the older algorithm interface, `init(m_ACTSGeoSvc().actsGeoProvider(), logger())` takes two inputs.  I need and use both inputs, can I revert to the older interface or is something that I am missing with newer interface?

ERROR:
======

`/home/runner/work/EICrecon/EICrecon/src/algorithms/tracking/SecondaryVertexFinder.h:60:8: error: 'eicrecon::SecondaryVertexFinder::init' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
  void init(std::shared_ptr<const ActsGeometryProvider> geo_svc,
       ^
/opt/local/include/algorithms/algorithm.h:87:16: note: hidden overloaded virtual function 'algorithms::Algorithm<algorithms::Input<edm4eic::ReconstructedParticleCollection, std::vector<const ActsExamples::Trajectories *>>, algorithms::Output<edm4eic::VertexCollection, edm4eic::VertexCollection>>::init' declared here: different number of parameters (0 vs 2)
  virtual void init() {}`

bishoyDHD committed:
Reverted back to  interface in .

bishoyDHD committed:
Reverted back to  interface in .

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

bishoyDHD committed:
Cleaned up iwyu

bishoyDHD committed:
Merge branch 'secvertexAMVF' of https://github.com/eic/EICrecon into displacedvertex

bishoyDHD committed:
Cleaning up iwyu from

bishoyDHD committed:
Update Sec vertex alorithm. Independent variables for primary and secondary vertex finding

bishoyDHD committed:
Made corrections to sec vtx finder to ensure that track container not empty. Avoid hard segfualts

bishoyDHD committed:
Cleaned up AMVF Primary and Sec. vertex

bishoyDHD committed:
Resolve some conflicts with the JEventProcessorPODIO.cc

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

bishoyDHD committed:
removed unused var

bishoyDHD committed:
removed unused var

bishoyDHD committed:
iwyu fix

bishoyDHD committed:
removed format include directive

bishoyDHD commented:
Alright folks. Please help me out here. This online precompile check keeps asking for me to `iwyu` by:
`#include <format>`

When I do, the precompiler cannot even find it. I had to upgrade my container in order to be able compile while having the `#include <format>`. I mean, the code is really ready to be reviewed and merged.

wdconinc commented:
> I had to upgrade my container in order to be able compile while having the `#include <format>`.

Yes, you should run with an up to date container.

> I mean, the code is really ready to be reviewed and merged.

Last time I looked, today afternoon, there was still no algorithms interface as requested.

bishoyDHD commented:
> > I had to upgrade my container in order to be able compile while having the `#include <format>`.
> 
> Yes, you should run with an up to date container.
> 
> > I mean, the code is really ready to be reviewed and merged.
> 
> Last time I looked, today afternoon, there was still no algorithms interface as requested.

Regarding the algorithms interface:
I have tested out the Algorithm interface last week. The problem, it adds a level of complexity that is not efficient for the secondary vertex setup. For example, I have functions that need to independently calculate the Primary and Secondary Vertex using the AMVF. Using the Algorithm::Input & Algorithm::Output means that I would have to unwrap those types and then pass them to the respective functions. That is just not efficient.

bishoyDHD committed:
Merge branch 'main' into secvertexAMVF

veprbl commented:
@wdconinc I know that @bishoyDHD did a good faith attempt at adapting his code to the algorithms interface. As you know, there are objective issues with interfacing `ACTS_LOCAL_LOGGER` with algorithms that are the same for existing CKFTracking code. The issues with passing code to functions I can't testify, but we probably can relatively easily address in further refactoring. My proposal is to review PR without asking for algorithms interface. Likely, it will be to either me or you to take on this technical debt.

wdconinc commented:
> @wdconinc I know that @bishoyDHD did a good faith attempt at adapting his code to the algorithms interface. As you know, there are objective issues with interfacing `ACTS_LOCAL_LOGGER` with algorithms that are the same for existing CKFTracking code. The issues with passing code to functions I can't testify, but we probably can relatively easily address in further refactoring. My proposal is to review PR without asking for algorithms interface. Likely, it will be to either me or you to take on this technical debt.

That's not the objection that's been raised. I'm ok with keeping an extra helper function that passes a spdlog logger from the factory to the algorithm, but there is no technical reason that the rest can't be using the algorithms interface.

wdconinc commented on src/algorithms/tracking/SecondaryVertexFinder.cc:
```diff
@@ -0,0 +1,419 @@
+}
+
+std::tuple<std::unique_ptr<edm4eic::VertexCollection>, std::unique_ptr<edm4eic::VertexCollection>>
+eicrecon::SecondaryVertexFinder::produce(
```
Please use the function signature `void process(const Input&, const Output&) const`. In particular, this function must be void, just be const, and should not create its own collections.

veprbl commented:
> > @wdconinc I know that @bishoyDHD did a good faith attempt at adapting his code to the algorithms interface. As you know, there are objective issues with interfacing `ACTS_LOCAL_LOGGER` with algorithms that are the same for existing CKFTracking code. The issues with passing code to functions I can't testify, but we probably can relatively easily address in further refactoring. My proposal is to review PR without asking for algorithms interface. Likely, it will be to either me or you to take on this technical debt.
> 
> That's not the objection that's been raised. I'm ok with keeping an extra helper function that passes a spdlog logger from the factory to the algorithm, but there is no technical reason that the rest can't be using the algorithms interface.

I wrote "passing code" instead of "passing pointers [to other methods of the algorithm]". My assumption is that it's a trivial change, but there is some C++ gotcha that tripped up @bishoyDHD in his attempt. I'd rather work with code that is final and effective before doing the refactoring we are talking about.

veprbl committed:
SecondaryVertexFinder: adopt algorithms convention for process call

veprbl committed:
SecondaryVertexFinder: use <> includes where appropriate

veprbl committed:
add missing copyright years

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

veprbl committed:
Update src/algorithms/tracking/TrackProjector.cc

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

veprbl committed:
Merge branch 'main' into secvertexAMVF

veprbl committed:
avoid extra small epsilon

veprbl committed:
remove comment that's already in Config

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

veprbl committed:
SecondaryVertexFinder: avoid warnings of init shadowing its proper definition

veprbl commented:
Looks like we have two comments unaddressed from the first round of review. One on using spdlog, which we might punt on. And another about units of initialVariances and moving that to config.

bishoyDHD committed:
Added initialVariances to Config

bishoyDHD committed:
Merge branch 'secvertexAMVF' of https://github.com/eic/EICrecon into displacedvertex

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

bishoyDHD commented:
> Looks like we have two comments unaddressed from the first round of review. One on using spdlog, which we might punt on. And another about units of initialVariances and moving that to config.

Moved to Config

bishoyDHD committed:
clean up iwyu

bishoyDHD committed:
Merge branch 'secvertexAMVF' of https://github.com/eic/EICrecon into displacedvertex

wdconinc commented on src/algorithms/tracking/SecondaryVertexFinder.cc:
```diff
@@ -0,0 +1,414 @@
+    });                              // vtxposition
+    eicvertex.setPositionError(cov); // covariance
+
+    for (const auto& t : vtx.tracks()) {
+      const auto& par = vertexfinderConfigSec.extractParameters(t.originalParams);
+      m_log->trace("Track local position from vertex = {} mm, {} mm",
+                   par.localPosition().x() / Acts::UnitConstants::mm,
+                   par.localPosition().y() / Acts::UnitConstants::mm);
+      float loc_a = par.localPosition().x();
+      float loc_b = par.localPosition().y();
+
+      for (const auto& part : reconParticles) {
+        const auto& tracks = part.getTracks();
+        for (const auto& trk : tracks) {
+          const auto& traj    = trk.getTrajectory();
+          const auto& trkPars = traj.getTrackParameters();
+          for (const auto& par : trkPars) {
+            double EPSILON = std::numeric_limits<double>::epsilon(); // mm
+            if (std::abs((par.getLoc().a / edm4eic::unit::mm) - (loc_a / Acts::UnitConstants::mm)) <
+                    EPSILON &&
+                std::abs((par.getLoc().b / edm4eic::unit::mm) - (loc_b / Acts::UnitConstants::mm)) <
+                    EPSILON) {
+              m_log->trace(
+                  "From ReconParticles, track local position [Loc a, Loc b] = {} mm, {} mm",
+                  par.getLoc().a / edm4eic::unit::mm, par.getLoc().b / edm4eic::unit::mm);
+              eicvertex.addToAssociatedParticles(part);
+            } // endif
+          } // end for par
+        } // end for trk
+      } // end for part
+    } // end for t
```
This whole matching is not ideal, and I'd like to see it removed instead of getting into our code base and left for others to refactor later.

The input to seed finding is (only) a set of track parameters. You have these track parameters as input in the form of ReconstructedParticles. But you ALSO pass and use the ActsExamples::Trajectories (which we are actively trying to get away from...). Then after seed finding you have to run this nested set of loops for matching with ReconstructedParticles. Why not start from ReconstructedParticles to fill your input tracks? Then when you get the vertices they will point to the tracks that are filled parallel to the ReconstructedParticles vector. There will be no matching loops necessary.

veprbl replied:
We told @bishoyDHD that copy-pasting the https://github.com/eic/EICrecon/blob/246dae58831f605639f8086abab1a0b450d093cb/src/algorithms/tracking/IterativeVertexFinder.cc#L155-L182 is an acceptable workaround to the issue of associating to PODIO. I don't think writing a routine to convert ReconstructedParticles to Trajectories is a useful exercise, if we plan to move to PODIO for Acts EDM anyway. Even once we have that, there will be some matching loop to have.

wdconinc replied:
> I don't think writing a routine to convert ReconstructedParticles to Trajectories is a useful exercise.

Not what I'm suggesting. Look at the code. Instead of getTrackParameters from Trajectories, you can getTrackParameters from ReconstructedParticles, instead of doing exactly that in the loop later.

veprbl replied:
Or the 3D position? But then we need to convert using bound surface.

wdconinc replied:
It's in the loop here in the segment that the comment is about:
```
   for (const auto& part : *reconParticles) { 
     const auto& tracks = part.getTracks(); 
     for (const auto& trk : tracks) { 
       const auto& traj    = trk.getTrajectory(); 
       const auto& trkPars = traj.getTrackParameters();
```
i.e. [edm4eic::ReconstructedParticle::getTracks](https://github.com/eic/EDM4eic/blob/main/edm4eic.yaml#L257), [edm4eic::Track::getTrajectory](https://github.com/eic/EDM4eic/blob/main/edm4eic.yaml#L468), [edm4eic::Trajectory::getTrackParameters](https://github.com/eic/EDM4eic/blob/main/edm4eic.yaml#L432)

wdconinc replied:
The idea is this: https://github.com/eic/EICrecon/pull/1992 (not even tested compilation, and has some remaining FIXMEs).

wdconinc replied:
> non-existent conversion routine

In the sense that CKFTracking does this already (minus surface lookup since it's a perigee that is re-created).

wdconinc reviewed (commented):
Two major remaining comments from my side at this point:
- We should move the creation and initialization of algorithmic structures that are event-independent into the init function, not recreate them for each event in the process function.
- We should not base this algorithm on the input from ActsExamples::Trajectories, but on the input from ReconstructedParticles. That will avoid the combinatorics of the matching loop. (Even if you don't do this, the matching loop should use an early break.)

veprbl committed:
Merge branch 'main' into secvertexAMVF

veprbl commented on src/algorithms/tracking/SecondaryVertexFinder.cc:
```diff
@@ -0,0 +1,414 @@
+  }
+
+  vertexfinderConfigSec.extractParameters.connect<&Acts::InputTrack::extractParameters>();
+  vertexFitterConfigSec.trackLinearizer.connect<&LinearizerSec::linearizeTrack>(&linearizerSec);
```
This is a use-after-move?

veprbl commented on src/algorithms/tracking/SecondaryVertexFinderConfig.h:
```diff
@@ -0,0 +1,39 @@
+struct SecondaryVertexFinderConfig {
+  int maxVertices        = 20;
+  float maxIterations    = 500;
+  float maxSecIterations = 1000;
```
```suggestion
  unsigned int maxVertices        = 20;
  unsigned int maxIterations    = 500;
  unsigned int maxSecIterations = 1000;
```

veprbl committed:
Update src/algorithms/tracking/SecondaryVertexFinderConfig.h

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

veprbl committed:
Update src/factories/tracking/SecondaryVertexFinder_factory.h

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

veprbl committed:
Change tracksMaxZinterval type from unsigned int to float

veprbl committed:
Update src/factories/tracking/SecondaryVertexFinder_factory.h

Co-authored-by: Wouter Deconinck <wdconinc@gmail.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

veprbl committed:
Update src/algorithms/tracking/SecondaryVertexFinder.cc

veprbl commented on src/algorithms/tracking/SecondaryVertexFinder.cc:
```diff
@@ -0,0 +1,414 @@
+  }
+
+  vertexfinderConfigSec.extractParameters.connect<&Acts::InputTrack::extractParameters>();
+  vertexfinderConfigSec.trackLinearizer.connect<&LinearizerSec::linearizeTrack>(&linearizerSec);
```
Turns out this was not a thing

I don't see this in Acts example
https://github.com/acts-project/acts/blob/a8e9f5cb396c27a999b5bc052635581bf46063fc/Examples/Algorithms/Vertexing/src/AdaptiveMultiVertexFinderAlgorithm.cpp#L187-L188

```suggestion
```

veprbl committed:
Apply suggestions from code review

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

veprbl committed:
Update src/algorithms/tracking/SecondaryVertexFinder.cc

wdconinc committed:
Secvertex amvf to be merged into main EICrecon repo (fix: iwyu) (#2187)

This PR applies the include-what-you-use fixes as suggested by
https://github.com/eic/EICrecon/actions/runs/19345829824.
Please merge this PR into the branch `secvertexAMVF`
to resolve failures in PR #1915.

Auto-generated by [create-pull-request][1]

[1]: https://github.com/peter-evans/create-pull-request

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

veprbl committed:
homogenize secondary and primary vertex finders

veprbl committed:
add copyright

veprbl committed:
Merge branch 'main' into secvertexAMVF

veprbl committed:
use collection names consistent with the IVF after cuts and Helix method

wdconinc committed:
Secvertex amvf to be merged into main EICrecon repo (fix: iwyu) (#2252)

This PR applies the include-what-you-use fixes as suggested by
https://github.com/eic/EICrecon/actions/runs/20286708727.
Please merge this PR into the branch `secvertexAMVF`
to resolve failures in PR #1915.

Auto-generated by [create-pull-request][1]

[1]: https://github.com/peter-evans/create-pull-request

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

veprbl commented:
We'll need to update this following
 - [x] #2224
 - [x] #2311
 - [x] #2317

veprbl committed:
Merge remote-tracking branch 'origin/main' into secvertexAMVF

 Conflicts:
	src/algorithms/tracking/TrackProjector.cc
	src/global/tracking/tracking.cc

veprbl committed:
fix: remove DD4hepBField dependency from SecondaryVertexFinder

Replace the removed eicrecon::BField::DD4hepBField with the standard
Acts::MagneticFieldProvider interface, matching the migration done in
commit 789bc2874b22 for CKFTracking and IterativeVertexFinder.

- Replace #include DD4hepBField.h with <Acts/MagneticField/MagneticFieldProvider.hpp>
- Change m_BField type to std::shared_ptr<const Acts::MagneticFieldProvider>
- Use m_geoSvc->getFieldProvider() directly instead of dynamic_pointer_cast
- Use Acts::MagneticFieldContext{} instead of BFieldVariant
- Remove #if Acts_VERSION_MAJOR guards for DD4hepBField casts
- Remove obsolete includes (detail/ContextType.hpp, fmt/core.h)

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

epic-capybara committed:
Secvertex amvf to be merged into main EICrecon repo (fix: iwyu) (#2626)

This PR applies the include-what-you-use fixes as suggested by
https://github.com/eic/EICrecon/actions/runs/24670790622.
Please merge this PR into the branch `secvertexAMVF`
to resolve failures in PR #1915.

Auto-generated by [create-pull-request][1]

[1]: https://github.com/peter-evans/create-pull-request

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

veprbl committed:
feat: replace ActsExamples::Trajectories with ConstTrackContainer in SecondaryVertexFinder

Migrate SecondaryVertexFinder from the removed ActsExamples::Trajectories
to ActsExamples::ConstTrackContainer, matching the migration done in
commit 528e76afc538 for IterativeVertexFinder and other tracking algorithms.

- Change Algorithm input type from vector<Trajectories> to ConstTrackContainer
- Replace trajectory tip iteration with ConstTrackContainer iteration,
  extracting BoundTrackParameters from track proxy
- Update factory to use Input<ConstTrackContainer> and pass the container
  pointer directly to process()
- Update tracking.cc wiring: CentralCKFActsTrajectories -> CentralCKFActsTracks
- Remove gsl/pointers dependency (no longer needed without gsl::not_null
  trajectory wrappers)

veprbl committed:
feat: replace ConstTrackContainer with separate TrackStates/Tracks in SecondaryVertexFinder

Migrate SecondaryVertexFinder from ActsExamples::ConstTrackContainer to
the split Acts::ConstVectorMultiTrajectory + Acts::ConstVectorTrackContainer
interface, matching the migration done in commit 033e22b8 for
IterativeVertexFinder and other tracking algorithms.

- Change Algorithm input type to accept TrackStates and Tracks separately
- Reconstruct ConstTrackContainer locally in each method from the two
  underlying containers
- Update factory to use two separate Input<> declarations with assertions
- Update tracking.cc wiring to pass CentralCKFActsTrackStates alongside
  CentralCKFActsTracks

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

epic-capybara committed:
Secvertex amvf to be merged into main EICrecon repo (fix: iwyu) (#2627)

This PR applies the include-what-you-use fixes as suggested by
https://github.com/eic/EICrecon/actions/runs/24674429934.
Please merge this PR into the branch `secvertexAMVF`
to resolve failures in PR #1915.

Auto-generated by [create-pull-request][1]

[1]: https://github.com/peter-evans/create-pull-request

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

veprbl committed:
SecondaryVertexFinderConfig: cool units

veprbl commented on src/algorithms/tracking/SecondaryVertexFinderConfig.h:
```diff
@@ -0,0 +1,39 @@
+  float maxDistToLinPoint          = 5.5 * Acts::UnitConstants::mm;
+  // Bin extent in z-direction
+  float spatialBinExtent         = 25 * Acts::UnitConstants::um;
+  Acts::Vector4 initialVariances = Acts::Vector4{1e+2, 1e+2, 1e+2, 1e+8};
```
Done.

veprbl committed:
add docstring

veprbl committed:
SecondaryVertexFinderConfig: more cool units

veprbl commented on src/algorithms/tracking/SecondaryVertexFinder.h:
```diff
@@ -0,0 +1,91 @@
+      : SecondaryVertexFinderAlgorithm{name,
+                                       {"inputReconstructedParticles", "inputActsTrajectories"},
+                                       {"outputPrimaryVertices", "outputSecondaryVertices"},
+                                       ""} {}
```
Not anymore. Thanks.

veprbl committed:
JEventProcessorPODIO: relocate collections

ruse-traveler reviewed (approved):
Following discussion with @marrbnl and @Simple-Shyam, this is ready to go! There are 2 items to follow-up on for the next maintainer:

- Deduplication of code between vertexing modes (not critical, but preferable) and integration of the Secondary Vertex EDM when ready (critical)
- Validation of algorithm performance, e.g. checking the kaon invariant mass distribution
